README.md

```diff
@@ -19,12 +19,351 @@ These dependencies need to be installed, in addition to ROS:
+
+        // Publish the text to be said.
+        talk::Text msg;
+        msg.text = desire.text();
+        msg.id = desire.id();
+        m_talkPublisher.publish(msg);
+    }
+
+private:
+    void talkDoneSubscriberCallback(const talk::Done::ConstPtr& msg)
+    {
+        // Remove the desire once the text is said.
+        if (msg->id == desireId())
+        {
+            m_desireSet->removeDesire(msg->id);
+        }
+    }
+};
+```
+
+### Using the Desire Set
+The following subsections show how to use the desire set. The DesireSet class is thread-safe, so it can be called by any thread.
+Once a change has been made to the desire set, the solver will be run to update the filter states.
+To prevent this behavior, a transaction can be created, so the solver will be run when the transaction is destroyed.
+
+#### Add a Desire
+```cpp
+auto desire = make_unique<TalkDesire>("The text to say"); // Create the desire.
+desireSet->addDesire(std::move(desire)); // Add the desire to the set.
+```
+
+If you don't need to have access to the desire instance, there is a simplified syntax.
+```cpp
+auto id = desireSet->addDesire<TalkDesire>("The text to say"); // Create the desire, add the desire to the set and return the id.
+```
+
+#### Remove a Desire
+```cpp
+desireSet->remove(id); // Remove the desire that has the provided id.
+```
+
+#### Remove All Desires
+```cpp
+desireSet->clear(); // Remove all desires
+```
+
+#### Remove All Desires of a Specific Type
+```cpp
+desireSet->removeAllDesiresOfType(DesireType::get<TalkDesire>()); // Remove all TalkDesire instances.
+```
+
+A simplied syntax exists.
+```cpp
+desireSet->removeAllDesiresOfType<TalkDesire>(); // Remove all TalkDesire instances.
+```
+
+#### Check Whether the Set Contains a Desire
+```cpp
+desireSet->contains(id); // Return a boolean indicating whether the set contains a desire that has the provided id.
+```
+
+#### Check Whether the Set Contains a Desire of a Specific Type
+```cpp
+// Return a boolean indicating whether the set contains a desire of type TalkDesire.
+desireSet->containsAnyDesiresOfType(DesireType::get<TalkDesire>());
+```
+
+A simplied syntax exists.
+```cpp
+// Return a boolean indicating whether the set contains a desire of type TalkDesire.
+desireSet->containsAnyDesiresOfType<TalkDesire>();
+```
+
+#### How to Create a Transaction
+The following example shows how to use transactions.
+
+```cpp
+{
+    auto transaction = desireSet->beginTransaction();
+
+    // Change the desire set here.
+}
+// When the transaction is destroyed, the changes will be applied.
+```
+
+#### How to Use the Observer Pattern
+The following example shows how to create and add an observer to the desire set.
+
+```cpp
+#include <hbba_lite/core/DesireSet.h>
+#include <iostream>
+
+class LogDesireSetObserver : public DesireSetObserver
+{
+public:
+    void onDesireSetChanged(const std::vector<std::unique_ptr<Desire>>& desires) override
+    {
+        for (auto& desire : desires)
```

Comment:
use a `const auto&` desire?

Reply:
It's already the case because desires is const, no?

Reply:
I don't think so. The vector `desires` is const and can't be modified, but the contained objects can still be modified.
But I just realised that the same thing happens with unique_ptr.
Adding `const auto&` as I suggested only makes the unique_ptr const and not the desire...
But still, it means that you can't reassign the ptr in the unique_ptr, which is something.

Reply:
It is not possible to do that. See : https://godbolt.org/z/rWhY8e4M3.